src/cgutils.cpp

```diff
@@ -988,13 +994,17 @@ static Type *julia_struct_to_llvm(jl_value_t *jt, bool *isboxed)
                 latypes.push_back(lty);
             }
             if (!isTuple) {
-                structdecl->setBody(latypes);
+                if (is_vecelement_type(jt))
+                    // VecElement type is unwrapped in LLVM
+                    jst->struct_decl = latypes[0];
+                else
+                    structdecl->setBody(latypes);
             }
             else {
                 if (isvector && lasttype != T_int1 && !type_is_ghost(lasttype)) {
                     // TODO: currently we get LLVM assertion failures for other vector sizes
-                    bool validVectorSize = (ntypes == 2 || ntypes == 4 || ntypes == 6);
-                    if (0 && lasttype->isSingleValueType() && !lasttype->isVectorTy() && validVectorSize) // currently disabled due to load/store alignment issues
+                    bool validVectorSize = (ntypes == 2 || ntypes == 4 || ntypes == 8 || ntypes == 16);
```

Comment:
Is there an inherent reason why we can't pass arbitrary length tuples as LLVM vectors?

As an example, something simple like:

```jl
immutable Vec{N, T}
    data::NTuple{N, VecElement{T}}
end

@generated function Base.(:+){N, T}(x::Vec{N, T}, y::Vec{N, T})
    exp = """
    %3 = fadd <$(N) x $(d[T])> %0, %1
    ret <$(N) x $(d[T])> %3
    """
    return quote
            Vec(Base.llvmcall($exp,
               NTuple{N, VecElement{T}},
               Tuple{NTuple{N,VecElement{T}},NTuple{N,VecElement{T}}},
               x.data, y.data))
    end
end
```

fails when the tuple gets passed as an array but works fine when it is passed as a vector.